StringOps speedups

isaacl opened the pull request:
- StringBuilder size hints
- contains override
- better updated impl

isaacl committed:
StringOps speedups

- StringBuilder size hints
- contains override
- better updated impl

retronym commented on src/library/scala/collection/StringOps.scala:
```diff
@@ -56,7 +56,7 @@ final class StringOps(private val s: String)
     */
   def map(f: Char => Char): String = {
-    val sb = new StringBuilder
+    val sb = new StringBuilder(length)
```
As another optimization, we should actually use a `java.lang.StringBuilder` here and in similar spots in this file.

isaacl replied:
I actually did that initially and then reverted thinking it was anti-scala, can add back

som-snytt replied:
It would be neat if the compiler did this automatically, for a local usage, to fall back from scala to java API if possible. There are cases (scala.js) where it's not clear what is portable or best. Also, if my handle were `isaacl`, I'd go with `iscala`.

retronym replied:
For now, let's add it back. We're stuck with the Scala version in `protected[this] def newSpecificBuilder`, but for these method-local versions I'd suggest avoiding the needless wrapper of `scala.StringBuilder`.

retronym commented on src/library/scala/collection/StringOps.scala:
```diff
@@ -56,7 +56,7 @@ final class StringOps(private val s: String)
   def map(f: Char => Char): String = {
-    val sb = new StringBuilder
+    val sb = new StringBuilder(length)
     for (ch <- s) sb += f(ch)
```
Another opportunity: use a `while` loop here rather than the `foreach`

isaacl replied:
yes certainly can be done if we're going for max speedup

retronym commented on src/library/scala/collection/StringOps.scala:
```diff
@@ -159,8 +159,20 @@ final class StringOps(private val s: String)
+   *  @return     `true` if this $coll has an element that is equal (as
+   *              determined by `==`) to `elem`, `false` otherwise.
+   */
+  override def contains[C >: Char](elem: C): Boolean = s.indexOf(elem) >= 0
```
I would expect this to be an overload, rather than an override.

```
override def contains(elem: Char): Boolean = s.indexOf(elem) >= 0
```

@szeiger WDYT?

retronym replied:
Hmmm, perhaps I've confused myself here. This looks good.

retronym replied:
But it needs to be:

```scala
  override def contains[C >: Char](elem: C): Boolean = elem match {
    case c: Char => s.indexOf(c) >= 0
    case _ => super.contains(elem)
  }

```

isaacl replied:
I'm confused as well. char equality means that nothing but a char will return true, right?

som-snytt replied:
I think you get the enriched `StringOps.indexOf` with generic `C`? not native string indexOf.

isaacl replied:
```
override def contains[C >: Char](elem: C): Boolean = elem match {
    case c: Char => s.indexOf(c) >= 0
    case _ => false
  }
```
could you do this?

som-snytt replied:
Probably. The old machinery uses apply/charAt and == so there's no way to stumble across an element that redefines equals. I haven't followed new collections yet.

retronym replied:
That looks like it would change `new StringOps("a").contains('a'.toInt)` to false. Currently it is true, due to cooperative equality:

```
scala> Int.box(1) == Char.box(1)
res1: Boolean = true

scala> "a".contains('a'.toInt)
res3: Boolean = true
```

som-snytt commented on src/library/scala/collection/StringOps.scala:
```diff
@@ -159,8 +159,20 @@ final class StringOps(private val s: String)
     *  @throws IndexOutOfBoundsException if `index` does not satisfy `0 <= index < length`.
     */
-  def updated(index: Int, elem: Char): String =
-    fromSpecificIterable(new View.Updated(this, index, elem)) // TODO optimize
```
Nice TODO! This is like the famous scene where indiana jones is confronted by the exotic whip-wielding menace, then he pulls out a familiar and reliable response that gets the job done.

isaacl committed:
More StringOps optimizations

isaacl committed:
Rewrite impl again...

Due to JEP 254, StringBuilder is better
than char[] for minor manipulation of an
existing String.

isaacl commented:
hmm, not sure why my last commit broke the build... Also on the subject of StringBuilder wrappers I notice that `IterableOnce::addString` operates with a wrapped StringBuilder, slowing down mkString.

SethTisue commented:
> not sure why my last commit broke the build

```
[test] [error] /home/jenkins/workspace/scala-2.13.x-validate-main@2/src/repl-frontend/scala/tools/nsc/interpreter/shell/Scripted.scala:242: method get() is defined twice;
[test] [error]   the conflicting method get() was defined at line 240:23
[test] [error]     @BeanProperty val names           = asList("scala")
[test] [error]                       ^
[test] [error] one error found
```

isaacl commented:
/rebuild

lrytz commented:
The failure is due to:
```
scala> "hi".capitalize
res7: String = ()
```
the cause is in method `updated`, due to the import change:
```
scala> new scala.StringBuilder("hi").setCharAt(0, 'H').toString
res8: String = Hi

scala> new java.lang.StringBuilder("hi").setCharAt(0, 'H').toString
res9: String = ()
```

isaacl commented:
ahh derp that method returns null, unlike the rest of the api. thanks

som-snytt commented:
It's `void` for symmetry with `getChars`, obviously. :confounded: 

I think you must squash for green. :vulcan_salute: Green squash is zucchini.

szeiger commented on src/library/scala/collection/StringOps.scala:
```diff
@@ -28,17 +27,17 @@ final class StringOps(private val s: String)
   override def view: StringView = StringView(s)
   protected[this] def coll: String = s
-  override def toSeq: immutable.Seq[Char] = fromIterable(toIterable)
+  override def toSeq: immutable.Seq[Char] = s.toCharArray
```
This relies on the deprecated implicit conversion that copies the array a 2nd time for the `ImmutableArray` after copying it from the String in `toCharArray`. Creating a `WrappedString` should be much more efficient.

isaacl replied:
sorry! I didn't see this message in time, looks like you're on top of it.

szeiger reviewed (commented):
Looks like a nice improvement. We should take this a step further and give `StringOps` the same treatment as `ArrayOps` to ensure that no inherited inefficient implementations remain (and to make it a pure value class to avoid boxing).

SethTisue commented:
welcome, @isaacl, and thank you!
